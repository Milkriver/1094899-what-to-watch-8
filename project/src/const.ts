export enum AppRoute {
  Main = '/',
  SignIn = '/login',
  MyList = '/mylist',
  Film = '/films/:id',
  AddReview = '/films/:id/review',
  Player = '/player/:id'
}

export enum AuthorizationStatus {
  Auth = 'AUTH',
  NoAuth = 'NO_AUTH',
  Unknown = 'UNKNOWN',
}

export enum APIRoute {
  Movies = '/films',
  SingleMovie = '/films/:id',
  SameGenreMovies = '/films/:id/similar',
  PromoMovie = '/promo',
  FavouriteMovies = '/favorite',
  StatusFavouriteMovie = '/favorite/: film_id/: status',
  Reviews = '/comments/:id',
  Login = '/login',
  Logout = '/logout',
}
