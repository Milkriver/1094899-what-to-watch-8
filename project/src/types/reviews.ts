export interface IReview {
    id: number,
    movieId: number,
    author: string,
    date: string,
    rating: number,
    text: string,
  }
