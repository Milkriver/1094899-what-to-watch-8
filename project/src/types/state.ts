import { AuthorizationStatus } from '../const';
import { IMovie } from './common';
import { IReviewResponse } from './reviews';

export interface IState {
  currentGenre: string
  movies: IMovie[]
  isDataLoaded: boolean
  authorizationStatus: AuthorizationStatus,
  movie: IMovie,
  sameMovies: IMovie[],
  reviews: IReviewResponse[],
  promoMovie: IMovie,
}
