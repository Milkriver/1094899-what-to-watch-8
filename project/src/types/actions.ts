import { AxiosInstance } from 'axios';
import { Action } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { IState } from './state';

export enum ActionType {
  ChangeGenre = 'movieList/ChangeGenre',
  LoadMovies = 'movieList/LoadMovies',
  LoadSingleMovie = 'movieCard/LoadSingleMovie',
  LoadActiveMovie = 'movieCard/LoadActiveMovie',
  LoadSameGenreMovies = 'movieCard/LoadSameGenreMovies',
  LoadPromoMovie = 'movieCard/LoadPromoMovie',
  LoadFavouriteMovies = 'userList/LoadFavouriteMovies',
  AddFavoriteMovie = 'userList/AddFavoriteMovie',
  RemoveFavoriteMovie = 'userList/RemoveFavoriteMovie',
  LoadReviews = 'movieCard/LoadReviews',
  AddReview = 'movieCard/AddReview',
  RequireAuthorization = 'user/requireAuthorization',
  RequireLogout = 'user/requireLogout',
  LoadUserData = 'user/loadUserData',
}

export type ThunkActionResult<R = Promise<void>> = ThunkAction<R, IState, AxiosInstance, Action>;

export type ThunkAppDispatch = ThunkDispatch<IState, AxiosInstance, Action>;
