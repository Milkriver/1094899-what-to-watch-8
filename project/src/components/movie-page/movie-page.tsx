import { useEffect, useState } from 'react';
import { connect, ConnectedProps } from 'react-redux';
import { Link } from 'react-router-dom';
import { useHistory } from 'react-router';
import { AppRoute, AuthorizationStatus } from '../../const';
import { fetchReviewsAction, fetchSameGenreMoviesAction, fetchSingleMovieAction } from '../../store/api-actions';
import { ThunkAppDispatch } from '../../types/actions';
import { IState } from '../../types/state';
import { Footer } from '../footer/footer';
import { Header } from '../header/header';
import { MovieCardElement } from '../movie-card-element/movie-card-element';
import { Tabs } from '../tabs/tabs';
import MyListButton from '../my-list-button/my-list-button';

const mapStateToProps = ({ movie, sameMovies, reviews, authorizationStatus }: IState) => ({
  movie,
  sameMovies,
  reviews,
  authorizationStatus,
});

const mapDispatchToProps = (dispatch: ThunkAppDispatch) => ({
  onFetchMovie(cardId: string) {
    dispatch(fetchSingleMovieAction(cardId));
  },
  onFetchSameGenreMovies(cardId: string) {
    dispatch(fetchSameGenreMoviesAction(cardId));
  },
  onFetchReviews(cardId: string) {
    dispatch(fetchReviewsAction(cardId));
  },
});

const connector = connect(mapStateToProps, mapDispatchToProps);
type PropsFromRedux = ConnectedProps<typeof connector>;


function MoviePage({ sameMovies, reviews, onFetchMovie, onFetchSameGenreMovies, onFetchReviews, movie: activeMovie, authorizationStatus }: PropsFromRedux): JSX.Element {
  const currentPageId = window.location.pathname.replace(AppRoute.Film.replace(':id', ''), '');
  const sameMovie = sameMovies.slice(0, 4);
  const [activeCardId, setActiveMovieCardId] = useState<number | undefined>();
  const handleActiveCard = (id: number | undefined) => { setActiveMovieCardId(id); };
  const history = useHistory();

  useEffect(() => {
    onFetchMovie(currentPageId);
    onFetchSameGenreMovies(currentPageId);
    onFetchReviews(currentPageId);
  }, [onFetchMovie, onFetchReviews, onFetchSameGenreMovies, currentPageId]);
  return (

    <>
      <section className="film-card film-card--full" style={{ backgroundColor: activeMovie.backgroundColor }}>
        <div className="film-card__hero">
          <div className="film-card__bg">
            <img src={activeMovie.backgroundImage} alt={activeMovie.name} />
          </div>
          <h1 className="visually-hidden">WTW</h1>
          <Header />
          <div className="film-card__wrap">
            <div className="film-card__desc">
              <h2 className="film-card__title">{activeMovie.name}</h2>
              <p className="film-card__meta">
                <span className="film-card__genre">{activeMovie.genre}</span>
                <span className="film-card__year">{activeMovie.released}</span>
              </p>

              <div className="film-card__buttons">
                <button className="btn btn--play film-card__button" type="button" onClick={() => history.push(AppRoute.Player.replace(':id', activeMovie.id.toString()))}>
                  <svg viewBox="0 0 19 19" width="19" height="19"><use xlinkHref="#play-s"></use></svg>
                  <span>Play</span>
                </button>
                {authorizationStatus === AuthorizationStatus.Auth
                  ?

                  <MyListButton currentMovie={activeMovie} />
                  : ''}

                {authorizationStatus === AuthorizationStatus.Auth
                  ? <Link to={AppRoute.AddReview.replace(':id', activeMovie.id.toString())} className="btn film-card__button">Add review</Link>
                  : ''}
              </div>
            </div>
          </div>
        </div>

        <div className="film-card__wrap film-card__translate-top">
          <div className="film-card__info">
            <div className="film-card__poster film-card__poster--big">
              <img src={activeMovie.posterImage} alt={activeMovie.name} width="218" height="327" />
            </div>
            <Tabs movie={activeMovie} reviews={reviews} />
          </div>
        </div>
      </section>

      <div className="page-content">
        <section className="catalog catalog--like-this">
          <h2 className="catalog__title">More like this</h2>

          <div className="catalog__films-list">
            {
              <>
                {sameMovie.map((singleCard) => (
                  <MovieCardElement
                    key={singleCard.id}
                    activeCardId={activeCardId}
                    filmName={singleCard.name}
                    id={singleCard.id}
                    onMouseOver={handleActiveCard}
                    videolink={singleCard.previewVideoLink}
                    posterSrc={singleCard.previewImage}
                  />
                ))}
              </>
            }
          </div>
        </section>
        <Footer />
      </div>
    </>

  );
}

export { MoviePage };
export default connector(MoviePage);
