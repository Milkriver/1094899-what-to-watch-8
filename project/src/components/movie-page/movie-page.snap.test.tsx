import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';
import { Provider } from 'react-redux';
import { IMovie } from '../../types/common';
import { MoviePage } from './movie-page';
import { AuthorizationStatus } from '../../const';
import { configureMockStore } from '@jedmao/redux-mock-store';

const MOCK_MOVIE_1: IMovie = { id: 1, name: '', posterImage: '', previewImage: '', backgroundImage: '', backgroundColor: '', videoLink: '', previewVideoLink: '', description: '', rating: 1, scoresCount: 1, director: '', starring: ['', ''], runTime: 1, genre: '', released: 1, isFavorite: true };
const MOCK_MOVIE_2: IMovie = { id: 2, name: '', posterImage: '', previewImage: '', backgroundImage: '', backgroundColor: '', videoLink: '', previewVideoLink: '', description: '', rating: 2, scoresCount: 2, director: '', starring: ['', ''], runTime: 1, genre: '', released: 2, isFavorite: false };
const MOCK_MOVIES_ARRAY: IMovie[] = [MOCK_MOVIE_2, MOCK_MOVIE_1];
const MOCK_REVIEWS = [{ id: 1, user: { id: 1, name: '' }, rating: 5, comment: '', date: '' }];

const mockFn = () => Promise.resolve();
describe('Component: MoviePage', () => {
  const mockStore = configureMockStore();

  const store = mockStore({
    MOVIES: {
      movie: {
        id: 1, name: '', posterImage: '', previewImage: '', backgroundImage: '',
        backgroundColor: '', videoLink: '', previewVideoLink: '', description: '', rating: 1, scoresCount: 1, director: '', starring: ['', ''],
        runTime: 1, genre: '', released: 1, isFavorite: true,
      },
    },
    USER: {},
    REVIEWS: {},
  });

  store.dispatch = jest.fn();

  test('should render correctly', () => {
    const history = createMemoryHistory();
    const page = (
      <Provider store={store}>
        <Router history={history}>
          <MoviePage
            sameMovies={MOCK_MOVIES_ARRAY}
            reviews={MOCK_REVIEWS}
            onFetchMovie={mockFn}
            onFetchSameGenreMovies={mockFn}
            onFetchReviews={mockFn}
            onFetchActiveMovie={mockFn}
            movie={MOCK_MOVIE_1}
            authorizationStatus={AuthorizationStatus.Auth}
          />;
        </Router>
      </Provider>
    );
    const { container } = render(page);
    expect(container).toMatchSnapshot();
  });
});
