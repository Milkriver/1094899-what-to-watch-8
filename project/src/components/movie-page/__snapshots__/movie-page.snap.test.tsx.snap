// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MoviePage should render correctly 1`] = `
<div>
  <section
    class="film-card film-card--full"
  >
    <div
      class="film-card__hero"
    >
      <div
        class="film-card__bg"
      >
        <img
          alt=""
          src=""
        />
      </div>
      <h1
        class="visually-hidden"
      >
        WTW
      </h1>
      <header
        class="page-header film-card__head"
      >
        <div
          class="logo"
        >
          <a
            class="logo__link"
            href="/"
          >
            <span
              class="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              class="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              class="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          class="user-block"
        >
          <a
            class="user-block__link"
            href="/login"
          >
            Sign in
          </a>
        </div>
      </header>
      <div
        class="film-card__wrap"
      >
        <div
          class="film-card__desc"
        >
          <h2
            class="film-card__title"
          />
          <p
            class="film-card__meta"
          >
            <span
              class="film-card__genre"
            />
            <span
              class="film-card__year"
            >
              1
            </span>
          </p>
          <div
            class="film-card__buttons"
          >
            <button
              class="btn btn--play film-card__button"
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlink:href="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              class="btn btn--list film-card__button"
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlink:href="#in-list"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              class="btn film-card__button"
              href="/films/1/review"
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      class="film-card__wrap film-card__translate-top"
    >
      <div
        class="film-card__info"
      >
        <div
          class="film-card__poster film-card__poster--big"
        >
          <img
            alt=""
            height="327"
            src=""
            width="218"
          />
        </div>
        <div
          class="film-card__desc"
        >
          <nav
            class="film-nav film-card__nav"
          >
            <ul
              class="film-nav__list"
            >
              <li
                class="film-nav__item film-nav__item--active"
              >
                <div
                  class="film-nav__link"
                >
                  Overview
                </div>
              </li>
              <li
                class="film-nav__item"
              >
                <div
                  class="film-nav__link"
                >
                  Details
                </div>
              </li>
              <li
                class="film-nav__item"
              >
                <div
                  class="film-nav__link"
                >
                  Reviews
                </div>
              </li>
            </ul>
          </nav>
          <div>
            <div
              class="film-rating"
            >
              <div
                class="film-rating__score"
              >
                1
              </div>
              <p
                class="film-rating__meta"
              >
                <span
                  class="film-rating__level"
                >
                  Bad
                </span>
                <span
                  class="film-rating__count"
                >
                  1
                </span>
              </p>
            </div>
            <div
              class="film-card__text"
            >
              <p />
              <p
                class="film-card__director"
              >
                <strong>
                  Director: 
                  
                </strong>
              </p>
              <p
                class="film-card__starring"
              >
                <strong>
                  Starring: 
                  
                  
                </strong>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
  <div
    class="page-content"
  >
    <section
      class="catalog catalog--like-this"
    >
      <h2
        class="catalog__title"
      >
        More like this
      </h2>
      <div
        class="catalog__films-list"
      >
        <article
          class="small-film-card catalog__films-card"
        >
          <div
            class="small-film-card__image"
          >
            <img
              alt=""
              height="175"
              src=""
              width="280"
            />
            ;
          </div>
          <h3
            class="small-film-card__title"
          >
            <div
              class="small-film-card__link"
            />
          </h3>
        </article>
        <article
          class="small-film-card catalog__films-card"
        >
          <div
            class="small-film-card__image"
          >
            <img
              alt=""
              height="175"
              src=""
              width="280"
            />
            ;
          </div>
          <h3
            class="small-film-card__title"
          >
            <div
              class="small-film-card__link"
            />
          </h3>
        </article>
      </div>
    </section>
    <footer
      class="page-footer"
    >
      <div
        class="logo"
      >
        <a
          class="logo__link logo__link--light"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>
  ;
</div>
`;
